src/index.ts

```diff
@@ -57,23 +53,136 @@ function mergeWithCustomize(options: ICustomizeOptions) {
   };
 }
 
-function customizeArray(rules: ICustomizeRules) {
+type Rules = { [s: string]: CustomizeRule | Rules };
+
+function customizeArray(rules: Rules) {
   return (a: any, b: any, key: Key) => {
-    const match = Object.keys(rules).find((rule) => wildcard(rule, key)) || "";
+    const matchedRule =
+      Object.keys(rules).find(rule => wildcard(rule, key)) || "";
 
-    switch (rules[match]) {
-      case CustomizeRule.Prepend:
-        return [...b, ...a];
-      case CustomizeRule.Replace:
-        return b;
-      case CustomizeRule.Append:
-      default:
-        return [...a, ...b];
+    if (matchedRule) {
+      switch (rules[matchedRule]) {
+        case CustomizeRule.Prepend:
+          return [...b, ...a];
+        case CustomizeRule.Replace:
+          return b;
+        case CustomizeRule.Append:
+        default:
+          return [...a, ...b];
+      }
     }
+
+    return mergeWithRules({ rules, key, a, b });
   };
 }
 
-function customizeObject(rules: ICustomizeRules) {
+function mergeWithRules({
+  rules,
+  key,
+  a,
+  b
+}: {
+  rules: Rules;
+  key: Key;
+  a: any;
+  b: any;
+}) {
+  let currentRule: CustomizeRule | Rules = rules;
+  key.split(".").forEach(k => {
+    currentRule = currentRule[k];
+  });
+
+  if (isPlainObject(currentRule)) {
+    return mergeWithRule({ currentRule, a, b });
+  }
+
+  return [];
+}
+
+function mergeWithRule({
+  currentRule,
+  a,
+  b
+}: {
+  currentRule: CustomizeRule | Rules;
+  a: any;
+  b: any;
+}) {
+  const bAllMatches: any[] = [];
+  const ret = a.map(ao => {
+    const ret = {};
+
+    const rulesToMatch: string[] = [];
+    const operations = {};
+    Object.entries(currentRule).forEach(([k, v]) => {
+      if (v === CustomizeRule.Match) {
+        rulesToMatch.push(k);
+      } else {
+        operations[k] = v;
+      }
+    });
+
+    const bMatches = b.filter(o => {
+      const matches = rulesToMatch.every(
+        rule => ao[rule].toString() === o[rule].toString()
```

Comment:
I tried testing this on a local project, and am getting this error: 

```
[app]/node_modules/webpack-merge/dist/index.js:122
            var matches = rulesToMatch.every(function (rule) { return ao[rule].toString() === o[rule].toString(); });
                                                                               ^

TypeError: Cannot read property 'toString' of undefined
    at [app]/node_modules/webpack-merge/dist/index.js:122:80
    at Array.every (<anonymous>)
    at [app]/node_modules/webpack-merge/dist/index.js:122:40
    at Array.filter (<anonymous>)
    at [app]/node_modules/webpack-merge/dist/index.js:121:26
    at Array.map (<anonymous>)
    at mergeWithRule ([app]/node_modules/webpack-merge/dist/index.js:108:17)
    at mergeWithRules ([app]/node_modules/webpack-merge/dist/index.js:101:16)
    at [app]/node_modules/webpack-merge/dist/index.js:90:16
    at _joinArrays ([app]/node_modules/webpack-merge/dist/join-arrays.js:44:50)
```

due to having a [`parser` config](https://webpack.js.org/configuration/module/#ruleparser) in my rules: 

```js
rules: [
  {
    parser: {
      system: false,
    },
  },
  {
    test: /\.(js|ts)x?$/,
    ...
  },
  {
    test: /\.css$/i,
    ...
  },
],
```

For testing, I removed that rule completely and then started getting the error below because of the `exclude: /node_modules/` in the next rule.

```
[app]/node_modules/webpack-merge/dist/index.js:110
    var ret = a.map(function (ao) {
                ^

TypeError: a.map is not a function
    at mergeWithRule ([app]/node_modules/webpack-merge/dist/index.js:110:17)
    at [app]/node_modules/webpack-merge/dist/index.js:161:30
    at Array.forEach (<anonymous>)
    at [app]/node_modules/webpack-merge/dist/index.js:130:28
    at Array.map (<anonymous>)
    at mergeWithRule ([app]/node_modules/webpack-merge/dist/index.js:110:17)
    at mergeWithRules ([app]/node_modules/webpack-merge/dist/index.js:102:16)
    at [app]/node_modules/webpack-merge/dist/index.js:90:16
    at _joinArrays ([app]/node_modules/webpack-merge/dist/join-arrays.js:44:50)
    at [app]/node_modules/webpack-merge/dist/merge-with.js:32:17
```


I imagine edge cases like this are the reason you've redone the API so this might not be something you want to handle. Just figured I might report it early in case it helps.

Reply:
Thanks for testing! Can you send me the exact merge you tried so I can convert that as a test case to support?

> I imagine edge cases like this are the reason you've redone the API so this might not be something you want to handle. Just figured I might report it early in case it helps.

Exactly. With this PR we'll get a lot of the power of the old API back while being explicit about it.

Reply:
Sure thing. I used the [syntax in the README](https://github.com/survivejs/webpack-merge/pull/150/files#diff-04c6e90faac2675aa89e2176d2eec7d8R197-R209), and the relevant parts of the configs are below.

<details>
    <summary>Base config</summary>

```js
rules: [
  {
    parser: {
      system: false,
    },
  },
  {
    test: /\.(js|ts)x?$/,
    exclude: /node_modules/,
    use: {
      loader: require.resolve("babel-loader", { paths: [__dirname] }),
    },
  },
  {
    test: /\.css$/i,
    include: [/node_modules/, /src/],
    use: [
      {
        loader: require.resolve("style-loader", { paths: [__dirname] }),
      },
      {
        loader: require.resolve("css-loader", { paths: [__dirname] }),
        options: {
          modules: false,
        },
      },
    ],
  },
],
```

</details>

<details>
    <summary>Consumer config</summary>

```js
rules: [
  {
    test: /\.html$/i,
    use: ["html-loader"],
  },
  {
    test: /\.css$/i,
    use: [
      require.resolve("style-loader", { paths: [__dirname] }),
      {
        loader: require.resolve("css-loader", { paths: [__dirname] }),
        options: {
          importLoaders: 1,
        },
      },
      {
        loader: "postcss-loader",
      },
    ],
  },
],
```

</details>

Reply:
Can you still paste the merge you tried?

Reply:
Is this what you're asking? I'm sorry if I'm misunderstanding. 

```js
const defaultConfig = {
    module: {
      rules: [
        {
          parser: {
            system: false,
          },
        },
        {
          test: /\.(js|ts)x?$/,
          exclude: /node_modules/,
          use: {
            loader: require.resolve("babel-loader", { paths: [__dirname] }),
          },
        },
        {
          test: /\.css$/i,
          include: [/node_modules/, /src/],
          use: [
            {
              loader: require.resolve("style-loader", { paths: [__dirname] }),
            },
            {
              loader: require.resolve("css-loader", { paths: [__dirname] }),
              options: {
                modules: false,
              },
            },
          ],
        },
      ],
    },
  }

  const localConfig = {
    module: {
      rules: [
        {
          test: /\.html$/i,
          use: ["html-loader"],
        },
        {
          test: /\.css$/i,
          use: [
            {
              loader: require.resolve("style-loader", { paths: [__dirname] }),
            },
            {
              loader: require.resolve("css-loader", { paths: [__dirname] }),
              options: {
                importLoaders: 1,
              },
            },
            {
              loader: "postcss-loader",
            },
          ],
        },
      ],
    }
  }

  const config = mergeWithCustomize({
    customizeArray: customizeArray({
      module: {
        rules: {
          test: "match",
          use: {
            loader: "match",
            options: "replace"
          }
        }
      }
    })
  })(defaultConfig, localConfig)
```

I also tried with `use: "replace"` in the `customizeArray` object since in my case I wouldn't mind completely overriding all the other loaders if the test regexes match, but got the same error.

Reply:
That's perfect. Thanks!

Reply:
@filoxo I pushed a fix for this particular case. Please give it a go.

Note that for the system to work as you expect, you should use the form `use: [{ loader: "html-loader" }]` for each loader. See the test I added for the complete solution.

Reply:
Just tested it out and its working like a charm! I think consistency in the loader syntax makes total sense. I did notice that in the merged object the `style-loader` rule now has an `options` key (which is `undefined` so it won't cause any issues as-is) while neither source object had options (see below). I imagine that's an artifact of the comparison logic. 

```js
{
  module: {
    rules: [
      { parser: { system: false } },
      {
        test: /\.(js|ts)x?$/,
        exclude: /node_modules/,
        use: {
          loader: '[app]/node_modules/babel-loader/lib/index.js'
        }
      },
      {
        test: /\.css$/i,
        include: [ /node_modules/, /src/ ],
        use: [
          {
            loader: '[app]/node_modules/style-loader/dist/cjs.js',
            options: undefined // TODO: this is weird, neither source object had an options value yet its key exists now
          },
          {
            loader: '[app]/node_modules/css-loader/dist/cjs.js',
            options: { importLoaders: 1 }
          },
          { loader: 'postcss-loader' }
        ]
      },
      { test: /\.html$/i, use: [ { loader: 'html-loader' } ] }
    ]
  },
}
```

I really appreciate the quick fix! Hopefully others benefit from this as well.

Reply:
Yeah, the options property is something I could likely remove from the result as well.

Reply:
@filoxo I fixed the `undefined` issue and also simplified the API a notch. Please have a look. Once you ok the feature, I'll publish. 👍

Reply:
Thanks for the update @bebraw. I tried out the new code and it seems to have reverted in behavior (despite that it passes in your local tests). I'm sort of stumped as to why that would be since my code hasn't changed. I threw together this example repo to show what I'm seeing since its a little more complex: https://github.com/filoxo/webpack-merge-test The instructions are included on how to run, and what should be expected. Let me know if you'd prefer something else to triage this.

Reply:
@filoxo I moved the behavior behind a new function, `mergeWithRules`. I see your code isn't using that yet.

Reply:
@filoxo Here's usage: https://github.com/survivejs/webpack-merge/pull/150/files#diff-04c6e90faac2675aa89e2176d2eec7d8R227 .

Reply:
🤦  I totally missed that, sorry! Just switched to `mergeWithRules` and it looks like its merged correctly now.

Reply:
No probs. I think I'll merge this tomorrow then and publish a new minor version with the feature. 👍